salt/modules/pkg_resource.py

```diff
@@ -311,22 +311,31 @@ def format_pkg_list(packages, versions_as_list, attr):
     '''
     ret = copy.deepcopy(packages)
     if attr:
+        ret_attr = {}
         requested_attr = set(['epoch', 'version', 'release', 'arch',
                               'install_date', 'install_date_time_t'])
 
         if attr != 'all':
-            requested_attr &= set(attr + ['version'])
+            requested_attr &= set(attr + ['version'] + ['arch'])
 
         for name in ret:
             versions = []
+            pkgname = name
             for all_attr in ret[name]:
                 filtered_attr = {}
                 for key in requested_attr:
                     if all_attr[key]:
                         filtered_attr[key] = all_attr[key]
                 versions.append(filtered_attr)
-            ret[name] = versions
-        return ret
+            try:
+                namepart, archpart = name.rsplit('.', 1)
+            except ValueError:
+                pass
+            else:
+                if archpart in [attrs['arch'] for attrs in versions]:
+                    pkgname = namepart
+            ret_attr.setdefault(pkgname, []).extend(versions)
+        return ret_attr
```

Comment:
This loop over `attrs` is not even needed. Try/except here is also overkill. And `pass` in try/except is just not nice. I would write it differently (and shorter): 

```python
try:
    _name, _arch = name.rsplit('.', 1)
except ValueError:
    _name, _arch = None, None

for all_attr in ret[name]:
    pkgname, filtered_attr = None, {}
    for key in requested_attr:
        if all_attr.get(key):
            filtered_attr[key] = all_attr[key]
    versions.append(filtered_attr)
    if _name and filtered_attr.get('arch', None) == _arch:
        pkgname = _name
    ret_attr.setdefault(pkgname or name, []).extend(versions)
```

This is also covering bug you've missed when `"foo."` is splitted as `['foo', '']`. If `attrs['arch']` is an empty string, you're in trouble. While it is untested, you've got the picture. :wink: